inject/src/main/java/io/micronaut/inject/ast/ClassElement.java

```diff
@@ -172,8 +209,117 @@ default boolean isProxy() {
      *
      * @return The default constructor if one is present
      */
-    default @NonNull Optional<MethodElement> getDefaultConstructor() {
-        return Optional.empty();
+    default Optional<MethodElement> getDefaultConstructor() {
+        Optional<MethodElement> staticCreator = findDefaultStaticCreator();
+        if (staticCreator.isPresent()) {
+            return staticCreator;
+        }
+        if (isInner() && !isStatic()) {
+            // only static inner classes can be constructed
+            return Optional.empty();
+        }
+        List<ConstructorElement> constructors = findConstructors()
+            .stream()
+            .filter(ctor -> ctor.getParameters().length == 0)
+            .collect(Collectors.toList());
+        if (constructors.isEmpty()) {
+            return Optional.empty();
+        }
+        if (constructors.size() == 1) {
+            return Optional.of(constructors.get(0));
+        }
+        return constructors.stream()
+            .filter(Element::isPublic)
+            .<MethodElement>map(c -> c)
+            .findFirst();
+
+    }
+
+    /**
+     * Find static creator.
```

Comment:
Clarify in the javadoc what a static creator is, as it is unclear for the user of this API